### ISSUE TITLE
Call function "Fatal error: Wrong number of parameters..."

### ISSUE THREAD

**Issue by KorsaR-ZN** [zephir-lang/zephir#938]
How to solve it?
# Zephir

```
Class Helper 
{
   public static function test(string! sp, string! b, string! c)
   {
        if func_num_args() > 3 {
          ....
        }

        return b . sp . c;
    }
}
```

**Outputs**

```
Helper::test(".", "1", "2"); // OK 
Helper::test(".", "1", "2", " ", "30"); // FAIL
```
# PHP

```
Class Helper 
{
   public static function test($sp, $b, $c)
   {
        if (func_num_args() > 3) {
          ....
        }

        return $b . $sp . $c;
    }
}
```

**Outputs**

```
Helper::test(".", 1, 2); // OK 
Helper::test(".", 1, 2, " ", 30); // OK 
```

**Comment by sergeyklay** [comment]
Yep, in php u can use syntax like this:

```php
class A
{
    function foo()
    {
        return func_get_args();
    }
}

/*
Array
(
    [0] => a
    [1] => b
    [2] => c
)
*/
```

**Comment by KorsaR-ZN** [comment]
@sergeyklay In Zephir this fatal error!

**Comment by sergeyklay** [comment]
So, if `Helper::test2("1", "2", "3", "4");` passed as workaround just do not declare args.

**Comment by KorsaR-ZN** [comment]
@sergeyklay I think this behavior is wrong for Zephir.
It should work in PHP or make the syntax for specifying a variable number of parameters. Example:
```
...

public static function test(string! sp, string! b, string! c, ...args) 
{
    ...
}

...
```
see http://php.net/manual/en/functions.arguments.php#functions.variable-arg-list.new

**Comment by sergeyklay** [comment]
+1 for variable number of parameters :)

**Comment by ovr** [comment]
It's caused by `zephir_fetch_parameters`
https://github.com/phalcon/zephir/blob/master/ext/kernel/main.c#L345

For example we can use it in `strict` mod and in not strict mod use similar function without this checks

cC @andresgutierrez

**Comment by nkt** [comment]
I suggest to mark this as `won't fix`, because there is no possible usage of this "missing" feature.
Also Zephir is not PHP, it's language for extensions, and this is standard behaviour for extension:
```
str_repeat('z', 2, 3, 4);
PHP Warning:  str_repeat() expects exactly 2 parameters, 4 given in php shell code on line 1
```

**Comment by KorsaR-ZN** [comment]
@nkt Your example gives a `warning` and in the case Zephir is `fatal error`

I think that if Zephir is a meta-language, it is possible to solve syntax and optimization at compile time, example:

**Before optimization before compiling**
```
....

public function some(string! a, string! b, int c, ...)

....
```

**After optimization**
```
....

public function some()
{
   var a, b, c;

   let a = func_get_arg(0),
       b = func_get_arg(1),
       c = (int) func_get_arg(2); // transparently convert

     if (!is_string(b) {
         // error
     }

     if (!is_string(a) {
         // error
     }

     if (!is_int(c) {
         // error
     }
    
   ....
}

....
```

**Comment by Jeckerson** [comment]
Fixed in `0.13.0` via `ZEND_PARSE_PARAMETERS`.
